packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,444 @@
+
+  @override
+  void initState() {
+    super.initState();
+    // TODO(jacobr): We shouldn't reach out to the WidgetsBinding.instance
+    // static here because we might not be in a tree that's attached to that
+    // binding. Instead, we should find a way to get to the PipelineOwner from
+    // the BuildContext.
+    _client = new _WidgetClient(WidgetsBinding.instance.pipelineOwner)
+      ..addListener(_update);
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  @override
+  void dispose() {
+    _client
+      ..removeListener(_update)
+      ..dispose();
+    WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
+  @override
+  void didChangeMetrics() {
+    setState(() {
+      // The root transform may have changed, we have to repaint.
+    });
+  }
+
+  void _update() {
+    SchedulerBinding.instance.addPostFrameCallback((Duration timeStamp) {
+      // We want the update to take effect next frame, so to make that
+      // explicit we call setState() in a post-frame callback.
+      if (mounted) {
+        // If we got disposed this frame, we will still get an update,
+        // because the inactive list is flushed after the widget updates
+        // are transmitted to the widget clients.
+        setState(() {
+          // The generation of the _WidgetInspectorListener has changed.
+        });
+      }
+    });
+  }
+
+  void _inspectAt(Offset position) {
+    if (!isSelectMode)
+      return;
+
+    final HitTestResult result = new HitTestResult();
+    final RenderIgnorePointer ignorePointer =
+        _ignorePointerKey.currentContext.findRenderObject();
+    final RenderBox userRender = ignorePointer.child;
+    // TODO
+    userRender.hitTest(result, position: position);
+
+    setState(() {
+      if (result.path.isNotEmpty) {
+        _selected = result.path
+            .firstWhere(((HitTestEntry t) => t.target is RenderBox))
+            .target;
+      } else {
+        _selected = null;
+      }
+    });
+  }
+
+  void _handlePointerDown(PointerDownEvent event) {
+    _lastPointerLocation = event.position;
+    _inspectAt(event.position);
+  }
+
+  void _handlePointerMove(PointerMoveEvent event) {
+    _lastPointerLocation = event.position;
+    _inspectAt(event.position);
+  }
+
+  void _handlePointerCancel(PointerCancelEvent event) {
+    setState(() {
+      isSelectMode = false;
+      _selected = null;
+    });
+  }
+
+  void _handlePointerUp(PointerUpEvent event) {
+    _lastPointerLocation = event.position;
+
+    // If the pointer up position is on the edge of the window assume that it
+    // indicates the pointer is being dragged off the edge of the display not
+    // a regular mouse up near the edge of the display. If the pointer is being
+    // dragged off the edge of the display we do not want to select anything.
+    final Rect bounds =
+        (Offset.zero & (ui.window.physicalSize / ui.window.devicePixelRatio))
+            .deflate(_kOffScreenMargin);
```

Comment:
Updated comment to explain that case.
    // A user can still select a widget that is only at the exact screen
    // margin by taping.